src/connection.c

```diff
@@ -160,12 +163,99 @@ static void handleConnected(const redisAsyncContext *c, int status)
         return;
     }
 
-    /* Call explicit connect callback (even if failed) */
+    /* Call callback even when failed */
     if (conn->connect_callback) {
         conn->connect_callback(conn);
     }
 }
 
+static void handleAuth(redisAsyncContext *c, void *r, void *privdata)
+{
+    Connection *conn = privdata;
+
+    redisReply *reply = r;
+    if (!reply) {
+        LOG_ERROR("Redis connection authentication failed: connection died");
+        ConnMarkDisconnected(conn);
+    } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG_ERROR("Redis connection authentication failed: %s", reply->str);
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else if (reply->type != REDIS_REPLY_STATUS || strcmp("OK", reply->str) != 0) {
+        LOG_ERROR("Redis connection authentication failed: unexpected response");
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else {
+        return connectionSuccess(conn);
+    }
+
+    return connectionFailure(conn);
+}
+
+/* Connection with authentication
+ *
+ * If the connection status is ok, will attempt to authenticate and only then call the connection callback
+ * If connection status is not ok, will immediately transition connection to CONN_CONNECT_ERROR
+ * User callback is called in all cases except when the connection has already been terminated externally
+ */
+
+static void handleConnectedWithAuth(Connection *conn, int status)
+{
+    if (status == REDIS_OK) {
+        /* don't try to authenticate if terminating */
+        if (conn->flags & CONN_TERMINATING) {
+            return connectionSuccess(conn);
+        }
+
+        if (redisAsyncCommand(ConnGetRedisCtx(conn), handleAuth, conn,
+                              "AUTH %s %s",
+                              conn->rr->config->cluster_user,
+                              conn->rr->config->cluster_password) != REDIS_OK) {
+            redisAsyncDisconnect(ConnGetRedisCtx(conn));
+            ConnMarkDisconnected(conn);
+            goto fail;
+        }
+
+        return;
+    }
+    /* if status != REDIS_OK, that means the connection failed, go direct to fail */
+fail:
+    connectionFailure(conn);
+}
+
+/* Connection without authentication
+ *
+ * Depending on status, state transitions to CONN_CONNECTED or CONN_CONNECT_ERROR.
+ * User callback is called in all cases except when the connection has already been terminated externally
+ */
+
+static void handleConnectedWithoutAuth(Connection *conn, int status) {
+    if (status == REDIS_OK) {
+        return connectionSuccess(conn);
+    } else {
+        return connectionFailure(conn);
```

Comment:
```suggestion
        connectionSuccess(conn);
    } else {
        connectionFailure(conn);
```

return with a void function is illegal according C as far as I know. We need to add -pedantic -Wextra warnings to compiler sometime(after cleaning up all warnings in the codebase), probably these options would warn us. 
```suggestion
        connectionSuccess(conn);
    } else {
        connectionFailure(conn);
```